rust/services/call/server_lib/src/proof/generator.rs

```
@@ -0,0 +1,236 @@
+use call_common::Metadata;
+use call_engine::{Call as EngineCall, CallGuestId};
+use call_host::{CycleEstimator, Host, PreflightResult, Prover, ProvingInput, Risc0CycleEstimator};
+use tracing::{error, info, instrument, warn};
+
+use crate::{
+    gas_meter::{Client as GasMeterClient, ComputationStage},
+    handlers::State as AppState,
+    metrics::Metrics,
+    preflight::{self},
+    proof::{
+        Error, allocate_error_to_state, preflight_error_to_state, set_metrics,
+        state::{State, set_state},
+        to_cycles, to_vgas,
+    },
+    proving::{self},
+    v_call::CallHash,
+};
+
+pub struct Generator {
+    gas_meter_client: Box<dyn GasMeterClient>,
+    vgas_limit: u64,
+    app_state: AppState,
+    call_hash: CallHash,
+    metrics: Metrics,
+}
+
+impl Generator {
+    pub fn new(
+        gas_meter_client: Box<dyn GasMeterClient>,
+        vgas_limit: u64,
+        app_state: AppState,
+        call_hash: CallHash,
+    ) -> Self {
+        Self {
+            gas_meter_client,
+            vgas_limit,
+            app_state,
+            call_hash,
+            metrics: Metrics::default(),
+        }
+    }
+
+    #[instrument(name = "proof", skip_all, fields(hash = %self.call_hash))]
+    pub async fn run(mut self, host: Host, call: EngineCall) {
+        info!("Generating proof");
+
+        match self.run_pipeline(host, call).await {
+            Ok(()) => {
+                info!("Proof generation completed successfully");
+            }
+            Err(_) => {
+                warn!("Proof generation failed");
+            }
```

Shouldn't we return the error to the caller of this method rather than swallow it and only log?

---

Previously, errors only updated the proof’s `State` and were not propagated back to the caller. The old `generate` function did not return anything (it had no return type).

Old `generate`:
```rs
pub async fn generate(
    call: EngineCall,
    host: Host,
    gas_meter_client: impl GasMeterClient,
    vgas_limit: u64,
    app_state: AppState,
    call_hash: CallHash,
) {
...
```